aptos-move/aptos-workspace-server/src/main.rs

```diff
@@ -1,277 +1,179 @@
-        faucet_health_checker.wait(None).await?;
-
-        println!(
-            "Faucet is ready. Endpoint: http://{}:{}",
-            IP_LOCAL_HOST, faucet_port
-        );
-
-        Ok(faucet_port)
-    };
-
-    (fut_faucet_port, fut_faucet_finish)
-}
-
-async fn start_all_services(test_dir: &Path) -> Result<()> {
-    // Step 1: spawn all services.
-    let (fut_node_api, fut_indexer_grpc, fut_node_finish) = start_node(test_dir)?;
+    // Phase 1: Start all services.
+    // Node
+    let (fut_node_api, fut_indexer_grpc, fut_node_finish) = services::node::start_node(test_dir)?;
 
     let fut_node_api = make_shared(fut_node_api);
     let fut_indexer_grpc = make_shared(fut_indexer_grpc);
-    let (fut_faucet, fut_faucet_finish) = start_faucet(
+
+    // Faucet
+    let (fut_faucet, fut_faucet_finish) = services::faucet::start_faucet(
         test_dir.to_owned(),
         fut_node_api.clone(),
         fut_indexer_grpc.clone(),
     );
 
-    let (res_node_api, res_indexer_grpc, res_faucet) =
-        tokio::join!(fut_node_api, fut_indexer_grpc, fut_faucet);
-
-    // Step 2: wait for all services to be up.
-    res_node_api
-        .map_err(anyhow::Error::msg)
-        .context("failed to start node api")?;
-    res_indexer_grpc
-        .map_err(anyhow::Error::msg)
-        .context("failed to start node api")?;
-    res_faucet.context("failed to start faucet")?;
+    // Docker Network
+    let docker_network_name = format!("aptos-workspace-{}", instance_id);
+    let (fut_docker_network, fut_docker_network_clean_up) =
+        create_docker_network(shutdown.clone(), docker_network_name);
+
+    // Indexer part 1: postgres db
+    let (fut_postgres, fut_postgres_finish, fut_postgres_clean_up) =
+        services::postgres::start_postgres(
+            shutdown.clone(),
+            fut_docker_network.clone(),
+            instance_id,
+        );
+    let fut_postgres = make_shared(fut_postgres);
 
-    println!(
-        "Indexer API is ready. Endpoint: http://{}:0/",
-        IP_LOCAL_HOST
+    // Indexer part 2: processors
+    let (fut_all_processors_ready, fut_any_processor_finish) = start_all_processors(
+        fut_node_api.clone(),
+        fut_indexer_grpc.clone(),
+        fut_postgres.clone(),
+    );
+    let fut_all_processors_ready = make_shared(fut_all_processors_ready);
+
+    // Indexer part 3: indexer API
+    let (fut_indexer_api, fut_indexer_api_finish, fut_indexer_api_clean_up) = start_indexer_api(
+        instance_id,
+        shutdown.clone(),
+        fut_docker_network.clone(),
+        fut_postgres.clone(),
+        fut_all_processors_ready.clone(),
     );
 
-    println!("ALL SERVICES STARTED SUCCESSFULLY");
-
-    // Step 3: wait for services to stop.
-    tokio::pin!(fut_node_finish);
-    tokio::pin!(fut_faucet_finish);
+    // Phase 2: Wait for all services to be up.
+    let all_services_up = async move {
+        tokio::try_join!(
```

Comment:
I suppose by design you can't configure as much as you can with localnet, e.g. using a host postgres / hasura, choosing to not run the faucet / particular processors, etc?

Reply:
My plan is to make these configurable at a later time. Design-wise it will require a bit more chore (storing things in `Option<impl Future>` or some other collections) but otherwise perfectly doable.

Reply:
> using a host postgres / hasura, choosing to not run the faucet / particular processors, etc?

These are crucial for the localnet unification